refactor: update `stats/incr/wmean` to handle nonnegative weights

anandkaranubc opened the pull request:
Resolves None

## Description

> What is the purpose of this pull request?

This pull request:

- Updates the implementation of `stats/incr/wmean` to correctly handle nonnegative weights (`w_i >= 0`) and ensures `mu` is updated only after encountering a non-zero weight.
- Adds corresponding test cases to validate the handling of zero weights.
- Updates existing test cases to correctly test the new outputs

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves None

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

- Now the implementation doesn't return `NaN` when weight is `NaN` or is not provided in the accumulator function. Instead it throws a TypeError.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

anandkaranubc committed:
refactor: update wmean implementation and tests to handle nonnegative weights

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

anandkaranubc committed:
refactor: improve error message

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

kgryte commented on lib/node_modules/@stdlib/stats/incr/wmean/lib/main.js:
```diff
@@ -107,6 +116,12 @@ function incrwmean() {
+		if ( !isNonNegativeNumber( w ) ) {
+			throw new TypeError( format( 'invalid argument. Must provide a nonnegative weight. Weight: `%s`.', w ) );
+		}
+		if (w === 0.0) {
```
```suggestion
		if ( w === 0.0 ) {
```

kgryte commented on lib/node_modules/@stdlib/stats/incr/wmean/lib/main.js:
```diff
@@ -107,6 +116,12 @@ function incrwmean() {
 			return mu;
 		}
+		if ( !isNonNegativeNumber( w ) ) {
+			throw new TypeError( format( 'invalid argument. Must provide a nonnegative weight. Weight: `%s`.', w ) );
```
```suggestion
			throw new TypeError( format( 'invalid argument. Second argument must be a nonnegative number. Value: `%s`.', w ) );
```

kgryte replied:
We prefer general error messages in order to minimize the total number of unique error messages in the code base. See `error/tools/*` for additional context.

anandkaranubc replied:
Thanks Athan! It will help me in the further implementations.

Applied the suggestions!

kgryte commented on lib/node_modules/@stdlib/stats/incr/wmean/lib/main.js:
```diff
@@ -97,7 +105,8 @@ function incrwmean() {
 	* @param {number} [x] - value
-	* @param {number} [w] - weight
+	* @param {NonNegativeNumber} [w] - weight
+	* @throws {TypeError} must provide a nonnegative number as the weight
```
```suggestion
	* @throws {TypeError} second argument must be a nonnegative number
```

kgryte commented on lib/node_modules/@stdlib/stats/incr/wmean/test/test.js:
```diff
@@ -46,6 +46,47 @@ tape( 'the initial accumulated value is `null`', function test( t ) {
+tape( 'the accumulator function ignores zero weights and only updates the mean after encountering a non-zero weight', function test( t ) {
+	var acc = incrwmean();
+	t.equal( acc(), null, 'returns expected value' );
+	t.equal( acc( 2.0, 0.0 ), null, 'returns the previous accumulated mean' );
+	t.equal( acc( 2.0, 2.0 ), 2.0, 'returns the current accumulated mean' );
+	t.equal( acc( 5.0, 0.0 ), 2.0, 'returns the previous accumulated mean' );
```
```suggestion
	t.equal( acc( 2.0, 0.0 ), null, 'returns expected value' );
	t.equal( acc( 2.0, 2.0 ), 2.0, 'returns expected value' );
	t.equal( acc( 5.0, 0.0 ), 2.0, 'returns expected value' );
```

kgryte commented on lib/node_modules/@stdlib/stats/incr/wmean/test/test.js:
```diff
@@ -46,6 +46,47 @@ tape( 'the initial accumulated value is `null`', function test( t ) {
+	t.end();
+});
+
+tape( 'the function throws an error if not provided a nonnegative number', function test( t ) {
```
```suggestion
tape( 'the accumulator function throws an error if provided a second argument which is not a nonnegative number', function test( t ) {
```

anandkaranubc committed:
refactor: apply suggestions from PR review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

anandkaranubc committed:
docs: update the ts index file

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

anandkaranubc commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into fix/wmean

kgryte commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into fix/wmean

Planeshifter commented on lib/node_modules/@stdlib/stats/incr/wmean/lib/main.js:
```diff
@@ -107,6 +116,12 @@ function incrwmean() {
 			}
 			return mu;
 		}
+		if ( !isNonNegativeNumber( w ) ) {
+			throw new TypeError( format( 'invalid argument. Second argument must be a nonnegative number. Value: `%s`.', w ) );
+		}
```
This throws when `w` is `undefined` (i.e., `acc(2.0)` with no weight), but the TypeScript type still declares `w` as optional (`w?: number`). Either the type needs to make `w` required, or the implementation needs to handle the missing-weight case without throwing.

Also, this now throws a `TypeError` for `NaN` weights — previously that case returned `NaN` (standard NaN poisoning). Throwing for `NaN` is a breaking change from the current behavior. Is that intentional?

Planeshifter commented on lib/node_modules/@stdlib/stats/incr/wmean/docs/types/index.d.ts:
```diff
@@ -28,7 +28,8 @@
 *
 * @param x - value
-* @param w - weight
+* @param {NonNegativeNumber} [w] - weight
+* @throws {TypeError} second argument must be a nonnegative number
```
Shouldn't have type definitions in `ts` files and we do not use bracket notation for optional parameters here:

```suggestion
* @param w - weight
* @throws second argument must be a nonnegative number
```

Planeshifter reviewed (changes requested):
Need to resolve discrepancy of weight `w` being required in implementation but marked as optional parameter.
